### PR TITLE
[fix](nereids)dphyper join reorder may cache wrong project list for project node

### PR DESCRIPTION
**Pull request by starocean999**
## Proposed changes

The top project of (T1, T2, T3) is different after reorder
we need merge **Project1** and **Project2** as **Project4** after reorder
```
T1 join T2 join T3:
   Project1(a, e + f)
       join(a = e)
           Project2(a, b + d as e)
               join(a = c)
                   T1(a, b)
                   T2(c, d)
       T3(e, f)
```

after reorder:
```
T1 join T3 join T2:
   Project4(a, b + d + f)
       join(a = c)
           Project3(a, b, f)
               join(a = e)
                   T1(a, b)
                   T3(e, f)
       T2(c, d)
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/joinorder/hypergraph/receiver/PlanReceiver.java [hunk 1]
```diff
@@ -335,58 +336,106 @@ private void makeLogicalExpression(Group root) {
...
+                    return getAlias == null ? e : getAlias.child();
+                });
+                return newInsideExpr == insideExpr ? expr
+                        : alias.withChildren(ImmutableList.of(newInsideExpr));
+            } else {
+                Alias getAlias = replaceMap.get(expr);
+                return getAlias == null ? expr : getAlias;
+            }
+        }).collect(ImmutableList.toImmutableList());
+    }
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/joinorder/hypergraph/receiver/PlanReceiver.java:380]: Already exist `mergeProjections(Project childProject)`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/joinorder/hypergraph/receiver/PlanReceiver.java [hunk 2]
```diff
@@ -335,58 +336,106 @@ private void makeLogicalExpression(Group root) {
...
-                        .collect(Collectors.toList());
-            }
-            if (projects.size() == 0) {
-                throw new RuntimeException("dphyer fail process project");
-            }
-            if (projects.size() == allProjects.size()) {
-                break;
-            }
         }
+        Preconditions.checkArgument(!projects.isEmpty() && projects.size() == allProjects.size());
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/joinorder/hypergraph/receiver/PlanReceiver.java:437]: checkStates

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Review by jackwener — APPROVED** [review]

**Review by englefly — APPROVED** [review]

**Pull request merged**